Respect CLJ_JVM_OPTS env var options when downloading clojure-tools

ikappaki opened the pull request:
Hi,

can you please consider patch to try and download clojure tools archive via a java subprocess that respects the `CLJ_JVM_OPTS` environment variable. This addresses #66.

I have disabled the babashka tests because it can't recognise the new imported `java.net.URLConnection` symbol, would it be possible to add it to the babashka arsenal please if review is successful please?

```
5:   (:import [java.lang ProcessBuilder$Redirect]
     ^--- Unable to resolve classname: java.net.URLConnection
```

Changes are:
1. The script will now attempt to download the clojure tools using the `java` executable first, fallback to direct download, and if that fails, display a message to the user where to copy the zip archive to. The java download methods requires Java11+/[JEP-330](https://openjdk.org/jeps/330).
2. If the archive was there though (say the user copied it over), the tool will not try to download it again, but unzip it.
3. An issue was identified in `unzip` with `Filesystems/newFilesystem` that was never closed (it is now in a `with-open` form).
4. An inconsistency was identified that `tools-dir` was passed to `clojure-tools-jar-download` as a destination directory instead of `libexec-dir`.
5. The `HttpURLConnection` is only set when downloading the clojure tools archive, if the src URL is `http`. For testing we use `file://` where possible. This introduced the `java.net.URLConnection` symbol, but unfortunately it is not available in babashka and thus the bb tests have been temporarily disabled until this is resolved.
6. Tests have been introduced to test all clojure tools download methodologies. A new dummy clojure tools archive is created on the fly when needed, and some java downloader tests only run when java > 11.
7. Updated `README.md` mentioned the new env variables and other minor corrections.

Thanks,

borkdude commented on src/borkdude/deps.clj:
```diff
@@ -225,43 +225,172 @@ For more info, see:
+  [dest-dir]
+  (let [dest-file (.getCanonicalPath (io/file dest-dir "ClojureToolsDownloader.java"))]
+    (spit dest-file
+          (str "/** Auto-generated by " *file* ". **/"
```
We can format this program by doing:

```
(str "<newline>




" ;; end of program
```

I it is more convenient to have it as a single string, so when we make changes, you don't have to re-format any other lines.

borkdude commented on src/borkdude/deps.clj:
```diff
@@ -225,43 +225,172 @@ For more info, see:
+    (when-not (.exists zip-file)
+      (warn "Downloading" ct-url-str "to" (str zip-file))
+      (.mkdirs dir)
+      (or (do (warn "\nAttempting download using " ::java-subprocess-method...
```
I think we can skip the new method when `CLJ_JVM_OPTS` hasn't been set

borkdude replied:
Should we probe the `java` version before trying the java method, like you're doing in the tests?

ikappaki replied:
The new method is now only called when `CLJ_JVM_OPTS` is set.

Unfortunately is not straightforward to check the java version number in the absence of `babashka.process`. This is because at least on Java 8 there is only a single `--version` cli option and this sends the output to `stderr`.  Thus `deps/shell-command` cannot be used to retrieve the version unless it is modified. 

Options:
1. Modify `shell-command` to accept an extra `:to-err-string` option to capture the `stderr` output of the subprocess. I can't see an easy way to differentiate between `stdout` (the current return value) and `stderr`, unless it returns a map?
2. Create a new fn to call on `ProcessBulder` to just pick up and return the java version from the err stream.

borkdude replied:
OK, I think we're going to let this slide, we will just demand Java 11 for this to work with CLJ_JVM_OPTS. Thank you.

borkdude commented on src/borkdude/deps.clj:
```diff
@@ -225,43 +225,172 @@ For more info, see:
+    (when-not (.exists zip-file)
+      (warn "Downloading" ct-url-str "to" (str zip-file))
+      (.mkdirs dir)
+      (or (do (warn "\nAttempting download using " ::java-subprocess-method...
```
I think the leading newline can be removed. The auto-resolved keyword  can be removed as well. 

We should only log once that we're attempting the download, regardless of the method, and not cause too much noise imo. 

We can log extra info when `DEPS_CLJ_DEBUG` has been set so when people have problems, we can ask them to set this variable.

ikappaki replied:
Done, only msgs before this patch are `warn'ed`, new simpified error messages are only printed out when the new `DEPS_CLJ_DEBUG` env variable is set.

borkdude commented on test/borkdude/deps_test.clj:
```diff
@@ -210,6 +211,144 @@
+          [temp-dir {}]
+          (let [dest-jar-file (fs/file temp-dir ct-jar-name)]
+            (with-redefs [deps/clojure-tools-download-direct
+                          (fn [& _] (throw (Exception. "Direct hould not be called.")))]
```
typo

ikappaki replied:
Corrected.

borkdude commented:
@ikappaki Beautiful PR, just a few minor comments.

For testing bb, I added `java.net.URLConnection` to bb now and you can get a dev build from:

https://github.com/babashka/babashka-dev-builds/releases/tag/v1.0.165-SNAPSHOT

You can install this with the installer script as well by providing a version with:

```
curl -sLO https://raw.githubusercontent.com/babashka/babashka/master/install
chmod +x ./install
./install --version 1.0.165-SNAPSHOT
```

Please update CI and turn bb tests back on.

borkdude commented on src/borkdude/deps.clj:
```diff
@@ -588,9 +601,10 @@ For more info, see:
-                                                                ["-Xms256m"])))
+           (let [java-clj-jvm-opts (when clj-jvm-opts (vec (concat clj-jvm-opts
+                                                                   proxy-settings
+                                                                   ["-Xms256m"])))]
```
What is the -Xms256m argument for?

ikappaki replied:
I picked it up from the `clj-main-cmd` a little bit further down the code, I assumed it was for some unknown to me use, shall I remove it?

ikappaki commented on test/borkdude/deps_test.clj:
```diff
@@ -210,6 +211,144 @@
+          [temp-dir {}]
+          (let [dest-jar-file (fs/file temp-dir ct-jar-name)]
+            (with-redefs [deps/clojure-tools-download-direct
+                          (fn [& _] (throw (Exception. "Direct hould not be called.")))]
```
Corrected.

ikappaki commented:
> @ikappaki Beautiful PR, just a few minor comments.
> 
> For testing bb, I added `java.net.URLConnection` to bb now and you can get a dev build from:
> 
> https://github.com/babashka/babashka-dev-builds/releases/tag/v1.0.165-SNAPSHOT
> 
> You can install this with the installer script as well by providing a version with:
> 
> ```
> curl -sLO https://raw.githubusercontent.com/babashka/babashka/master/install
> chmod +x ./install
> ./install --version 1.0.165-SNAPSHOT
> ```
> 
> Please update CI and turn bb tests back on.

Hi @borkdude,

I've addressed most of the review comments and updated the CI configs to use the babashka snapshot (markes as REVIEW) so as to run the `babashka-test` task. I'm ready to revert back to the normal babashka CI installation when you are ready to publish a new version.

In addition I've added some comments to the `-main` docstring about the use of the new variables, if you don't mind me intruding.

The only thing I haven't address is determining the java version, can you have a look at my reply and available options please?

Thanks,

borkdude commented:
I think this looks good. I looked into the Xms256 thing and really can't find where this came from. This is the initial commit of the deps.clj script and it was already there. I think I'll just remove it since it isn't there in the clojure script either.

borkdude commented:
Much thanks!
